src/marqo/tensor_search/tensor_search.py

```diff
@@ -2648,3 +2657,33 @@ def delete_documents(config: Config, index_name: str, doc_ids: List[str]):
             document_ids=doc_ids,
         )
     )
+
+def normalize_vector(embeddings: Union[List[List[float]], ndarray]) -> List[List[float]]:
+    """
+    Normalizes a list of vectors (embeddings) to have unit length.
+
+    Args:
+        embeddings (Union[List[List[float]], ndarray]): A list of vectors or a numpy ndarray of vectors to be normalized.
+
+    Returns:
+        List[List[float]]: A list of normalized vectors.
+    """
+
+    # Convert the input embeddings to a numpy array
+    if embeddings.__class__ == ndarray:
+        embeddings_array = embeddings
+    else:
+        embeddings_array = np.array(embeddings)
+
+    # Calculate the magnitude (Euclidean norm) of each vector along the last axis
+    magnitude = np.linalg.norm(embeddings_array, axis = -1, keepdims=True)
+
+    # Normalize each vector by dividing by its magnitude, handle zero magnitude case
+    if magnitude != 0:
+        embeddings_array = embeddings_array / magnitude
+    else:
+        embeddings_array = embeddings_array # Handle the zero vector case
```

Comment:
This looks like it should be a ValueError here and a 400 in the API response. Do take care not to fail the full add docs batch, but only the individual document with this vector

Reply:
Our earlier logic was not to raise a ValueError/400 in case magnitude was zero: https://github.com/marqo-ai/marqo/blob/mainline/src/marqo/tensor_search/tensor_search.py#L1908-L1911, are we sure we want to throw an error here?

Reply:
But also https://github.com/marqo-ai/marqo/blob/mainline/src/marqo/core/utils/vector_interpolation.py#L171

Reply:
Added logic to throw a new ZeroMagnitudeVectorError: https://github.com/marqo-ai/marqo/pull/970#diff-bc6ddb28e3809766b86804afac01ac632de1e691377c5d3f43d0333b97ba79faR303
https://github.com/marqo-ai/marqo/pull/970#diff-bc6ddb28e3809766b86804afac01ac632de1e691377c5d3f43d0333b97ba79faR848
https://github.com/marqo-ai/marqo/pull/970#diff-bc6ddb28e3809766b86804afac01ac632de1e691377c5d3f43d0333b97ba79faR1950